hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -694,7 +700,7 @@ private void refreshStoreFilesInternal(Collection<StoreFileInfo> newFiles) throw
     refreshStoreSizeAndTotalBytes();
   }
 
-  protected HStoreFile createStoreFileAndReader(final Path p) throws IOException {
+  public HStoreFile createStoreFileAndReader(final Path p) throws IOException {
```

Comment:
Who would be calling this outside of this package?

Reply:
Prior tho this, the "commit compaction" logic was implemented straight inside `HStore.doCompaction -> moveCompactedFilesIntoPlace -> moveFileIntoPlace` which was where `createStoreFileAndReader` was getting called. I believe this is "compaction specific" logic, and if we are aiming for pluggable compaction behaviour, it shouldn't be implemented in the HStore. That's why here we changed it to be `HStore.doCompaction -> storeEngine.compactor.commitCompaction`, so we had to eventually call it from the compactors.

Reply:
> Although, I also see that HStore#openStoreFiles is using one instance of createStoreFileAndReader, so we couldn't completely move this out of HStore. Is it better to lift one method into Store and leave other instances "internal" to HStore? That might help the smell a little, but an obvious solution isn't jumping out at me.

Yes, there are many other usages, pretty much everything that needs access to hfiles would end up accessing this method (indirectly). 

In terms of responsibility, it does seem right to me to have HStore the central point for store file accesses. Moving it to Store would still require some code duplication there, or exposing the two methods, which is what we are trying to avoid here.

Maybe a better solution is to define a lambda function in Compactor.commitCompaction, so that HStore can wrap up createStoreFileAndReader when calling Compactor.commitCompaction().

Reply:
> Maybe a better solution is to define a lambda function in Compactor.commitCompaction, so that HStore can wrap up createStoreFileAndReader when calling Compactor.commitCompaction().

Sounds resonable!